pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,423 @@
+// and once to the SQL trace on the leaseholder.
+// TODO(baptist): Look at logging traces on followers as well and sending back
+// to the leader in the MsgAppResp. It would need to be best effort, but might
+// still be useful.
+type RaftTracer struct {
+	// m is a map of all the currently traced entries for this replica. The
+	// aggregate size of the map across all replicas is equal to or less than
+	// numRegisteredStore unless the setting changes in which case we flush all
+	// entries on the next register call. We add to numRegistered before we
+	// update m, and delete from m before we remove from numRegistered to keep
+	// this invariant. On a setting change we flush all existing traces on the
+	// next call to register.
+	m syncutil.Map[kvpb.RaftIndex, traceValue]
+
+	// numRegisteredStore is the number of currently registered traces for this
+	// store, not this replica. The number of registered will always be less
+	// than the MaxConcurrentRaftTraces setting. If the setting is lowered, we
+	// flush all traces on all replicas.
+	numRegisteredStore *atomic.Int64
+
+	// numRegisteredReplica is the number of currently registered traces for
+	// this replica. The sum(numRegisteredReplica) <= numRegisteredStore. We set
+	// numRegisteredReplica to MaxInt64 when we close the tracer to prevent new
+	// registrations.
+	numRegisteredReplica atomic.Int64
+
+	// ctx is the ambient context for the replica and is used for remote
+	// traces. It contains the replica/range information. On each trace we
+	// additionally append the unique trace/span IDs.
+	ctx context.Context
+	st  *cluster.Settings
+}
+
+// NewRaftTracer creates a new RaftTracer with the given ambient context for the
+// replica.
+func NewRaftTracer(
+	ctx context.Context, st *cluster.Settings, numRegisteredStore *atomic.Int64,
+) *RaftTracer {
+	return &RaftTracer{ctx: ctx, st: st, numRegisteredStore: numRegisteredStore}
+}
+
+// MaxConcurrentRaftTraces is the maximum number of entries that can be traced
+// at any time on this store. Additional traces will be ignored until the number
+// of traces drops below the limit. Having too many active traces can negatively
+// impact performance as we iterate over all of them for some messages.
+//
+// TODO(baptist): Bump the default to a reasonable value like 10 that balances
+// usefulness with performance impact once we have validated the performance
+// impact.
+var MaxConcurrentRaftTraces = settings.RegisterIntSetting(
+	settings.SystemVisible,
+	"kv.raft.max_concurrent_traces",
+	"the maximum number of tracked raft traces",
+	0,
+	settings.NonNegativeInt,
+)
+
+// reserveSpace checks if should register a new trace. If there are too many
+// registered traces it will not register and return false. The soft invariant
+// is that numRegisteredStore <= numAllowed which can be temporarily violated if
+// the number .  This method will return true if we can add one to the number
```

Comment:
Could you fix the comment? Looks unfinished.

Reply:
Sorry - fixed!